File: cli/src/transport.rs
//! Thin USB transport for the rustprobe config protocol.
//!
//! Config commands ride on the vendor-bulk CMSIS-DAP v2 interfaces: one
//! request packet OUT (≤64 bytes), one response packet IN (≤64 bytes). This
//! module handles discovery and the single-packet transceive; the protocol
//! framing lives in [`crate::session`].

use anyhow::{bail, Context, Result};
use futures_lite::future::block_on;
use nusb::transfer::{Direction, EndpointType, RequestBuffer};
use nusb::{DeviceInfo, Interface, InterfaceInfo};

/// Rustprobe USB vendor id (shared with the Raspberry Pi debugprobe).
pub const VID: u16 = 0x2E8A;
/// Rustprobe USB product id.
pub const PID: u16 = 0x000C;

/// Bulk packet size (full-speed vendor bulk MPS).
const PACKET_SIZE: usize = 64;

/// Substring identifying a CMSIS-DAP interface in its interface string.
const DAP_MARKER: &str = "CMSIS-DAP";

/// True if `iface` is one of the probe's CMSIS-DAP interfaces.
pub fn is_dap_interface(iface: &InterfaceInfo) -> bool {
    iface
        .interface_string()
        .is_some_and(|s| s.contains(DAP_MARKER))
}

/// Enumerate attached rustprobe devices (matched by VID:PID).
pub fn list_devices() -> Result<Vec<DeviceInfo>> {
    Ok(nusb::list_devices()
        .context("enumerate USB devices")?
        .filter(|d| d.vendor_id() == VID && d.product_id() == PID)
        .collect())
}

/// Select a single device, optionally filtered by serial number.
pub fn find_device(serial: Option<&str>) -> Result<DeviceInfo> {
    let mut devices = list_devices()?;
    match serial {
        Some(s) => devices
            .into_iter()
            .find(|d| d.serial_number() == Some(s))
            .with_context(|| format!("no rustprobe with serial {s:?}")),
        None => match devices.len() {
            0 => bail!("no rustprobe devices found (VID:PID {VID:04X}:{PID:04X})"),
            1 => Ok(devices.remove(0)),
            n => bail!("{n} rustprobe devices attached; select one with --serial"),
        },
    }
}

/// A summary of a discovered device for `list`.
pub struct DeviceSummary {
    /// Device serial number ("<uid>:0"), if reported.
    pub serial: Option<String>,
    /// USB product string.
    pub product: Option<String>,
    /// Interface numbers whose string marks them as CMSIS-DAP.
    pub dap_interfaces: Vec<u8>,
}

/// Summarize a device without opening it (Linux exposes interface strings
/// through sysfs, so no control transfer is needed).
pub fn summarize(info: &DeviceInfo) -> DeviceSummary {
    DeviceSummary {
        serial: info.serial_number().map(str::to_owned),
        product: info.product_string().map(str::to_owned),
        dap_interfaces: info
            .interfaces()
            .filter(|i| is_dap_interface(i))
            .map(|i| i.interface_number())
            .collect(),
    }
}

/// An opened probe interface, ready to exchange config packets.
pub struct Probe {
    interface: Interface,
    ep_out: u8,
    ep_in: u8,
}

impl Probe {
    /// Open `info` and claim a DAP interface for config commands.
    ///
    /// Config commands are served on any probe interface, so the
    /// lowest-numbered CMSIS-DAP interface is used.
    pub fn open(info: &DeviceInfo) -> Result<Probe> {
        let iface_num = info
            .interfaces()
            .filter(|i| is_dap_interface(i))
            .map(|i| i.interface_number())
            .min()
            .context("device exposes no CMSIS-DAP interface")?;
        let device = info.open().context("open USB device")?;
        let interface = device
            .claim_interface(iface_num)
            .with_context(|| format!("claim interface {iface_num}"))?;
        let (ep_out, ep_in) =
            bulk_endpoints(&interface).context("locate bulk endpoints on DAP interface")?;
        Ok(Probe { interface, ep_out, ep_in })
    }

    /// Send one request packet and return the single response packet.
    pub fn transceive(&self, request: &[u8]) -> Result<Vec<u8>> {
        block_on(self.interface.bulk_out(self.ep_out, request.to_vec()))
            .into_result()
            .context("bulk OUT")?;
        let response = block_on(
            self.interface
                .bulk_in(self.ep_in, RequestBuffer::new(PACKET_SIZE)),
        )
        .into_result()
        .context("bulk IN")?;
        Ok(response)
    }
}

/// Find the (bulk OUT, bulk IN) endpoint addresses on the interface's first
/// alternate setting.
fn bulk_endpoints(interface: &Interface) -> Result<(u8, u8)> {
    let alt = interface
        .descriptors()
        .next()
        .context("interface has no alternate setting")?;
    let mut ep_out = None;
    let mut ep_in = None;
    for ep in alt.endpoints() {
        if ep.transfer_type() != EndpointType::Bulk {
            continue;
        }
        match ep.direction() {
            Direction::Out => ep_out.get_or_insert(ep.address()),
            Direction::In => ep_in.get_or_insert(ep.address()),
        };
    }
    match (ep_out, ep_in) {
        (Some(out), Some(inn)) => Ok((out, inn)),
        _ => bail!("interface lacks a bulk IN/OUT endpoint pair"),
    }
}

File: webui/site/js/transport.js
// WebUSB transport for the rustprobe config protocol.
//
// Mirrors cli/src/transport.rs: config commands ride on the vendor-bulk
// CMSIS-DAP v2 interfaces (class FF/00/00, interface string "CMSIS-DAP");
// one ≤64-byte packet OUT, one ≤64-byte packet IN. The pico-sdk reset
// interface is FF/00/01, so matching protocol 0 excludes it.

export const VID = 0x2e8a;
export const PID = 0x000c;
const PACKET_SIZE = 64;

export function webusbAvailable() {
  return typeof navigator !== "undefined" && !!navigator.usb;
}

/// Prompt the user to pick a rustprobe. Must be called from a user gesture.
export async function requestProbe() {
  return navigator.usb.requestDevice({
    filters: [{ vendorId: VID, productId: PID }],
  });
}

/// Previously-authorized rustprobes (no prompt needed).
export async function authorizedProbes() {
  const devices = await navigator.usb.getDevices();
  return devices.filter((d) => d.vendorId === VID && d.productId === PID);
}

function isDapInterface(iface) {
  const alt = iface.alternates[0];
  if (!alt || alt.interfaceClass !== 0xff || alt.interfaceSubclass !== 0x00 ||
      alt.interfaceProtocol !== 0x00) {
    return false;
  }
  // The interface string clinches it when the browser exposes it.
  if (alt.interfaceName && !alt.interfaceName.includes("CMSIS-DAP")) return false;
  return true;
}

/// Open `device` and claim its lowest-numbered CMSIS-DAP interface.
/// Returns a transport with a single-packet `transceive`.
export async function openProbe(device, log = () => {}) {
  await device.open();
  if (device.configuration === null) await device.selectConfiguration(1);

  const iface = device.configuration.interfaces
    .filter(isDapInterface)
    .sort((a, b) => a.interfaceNumber - b.interfaceNumber)[0];
  if (!iface) {
    await device.close();
    throw new Error("device exposes no CMSIS-DAP interface");
  }
  await device.claimInterface(iface.interfaceNumber);

  const eps = iface.alternates[0].endpoints.filter((e) => e.type === "bulk");
  const epOut = eps.find((e) => e.direction === "out")?.endpointNumber;
  const epIn = eps.find((e) => e.direction === "in")?.endpointNumber;
  if (epOut === undefined || epIn === undefined) {
    await device.close();
    throw new Error("DAP interface lacks a bulk IN/OUT endpoint pair");
  }

  return {
    device,
    interfaceNumber: iface.interfaceNumber,

    async transceive(request) {
      log(`→ ${hex(request)}`);
      const out = await device.transferOut(epOut, request);
      if (out.status !== "ok") throw new Error(`bulk OUT failed: ${out.status}`);
      const inn = await device.transferIn(epIn, PACKET_SIZE);
      if (inn.status !== "ok") throw new Error(`bulk IN failed: ${inn.status}`);
      const resp = new Uint8Array(inn.data.buffer, inn.data.byteOffset, inn.data.byteLength);
      log(`← ${hex(resp)}`);
      return resp;
    },

    async close() {
      try { await device.close(); } catch { /* already gone */ }
    },
  };
}

export function hex(bytes) {
  return Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join(" ");
}
